server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -709,11 +709,15 @@ public void onFailure(Exception e) {
 
         if (indexSettings.isSoftDeleteEnabled() && useRetentionLeasesInPeerRecovery == false) {
             final RetentionLeases retentionLeases = replicationTracker.getRetentionLeases();
+
             final Set<ShardRouting> shardRoutings = Sets.newHashSetWithExpectedSize(routingTable.size());
             for (int copy = 0; copy < routingTable.size(); copy++) {
-                shardRoutings.add(routingTable.shard(copy));
+                ShardRouting shardRouting = routingTable.shard(copy);
+                if (shardRouting.isPromotableToPrimary()) {
+                    shardRoutings.add(shardRouting);
```

Comment:
Ah no unfortunately this doesn't include the relocation targets, we need to handle them like this:

```suggestion
                    shardRoutings.add(shardRouting);
                    if (shardRouting.relocating()) {
                        shardRoutings.add(shardRouting.getTargetRelocatingShard());
                    }
```

I'm not sure we even need to build this set, we could just check that each promotable shard copy is assigned and has a tracked retention lease directly.